parlai/tasks/light_genderation_bias/agents.py

```diff
@@ -0,0 +1,405 @@
+        self.gender_dict = read_gender_tsv(os.path.join(_path(opt), GENDERED_LIST))
+        self.fixed_random = random.Random(42)
+
+        self.add_conditional = opt['add_conditional']
+        self.add_counterfactual = opt['add_counterfactual']
+        self.add_new_data = opt['add_new_data']
+
+        self.force_conditional = opt['force_conditional']
+        self.bucket_only = opt['bucket_only']
+        if self.force_conditional == 'none':
+            self.force_conditional = None
+        if self.bucket_only == 'none':
+            self.bucket_only = None
+
+        if shared and 'data' in shared:
+            self.data = shared['data']
+        else:
+            self.data = self._setup_data(opt)
+
+        super().__init__(opt, shared)
+        self.reset()
+
+    def num_episodes(self):
+        return len(self.data)
+
+    def num_examples(self):
+        return len(self.data)
+
+    def get_bucket(self, text):
+        """
+        Get the bin or "bucket" that a string falls into.
+
+        The bucket is determined based on the presence of male or femael gendered words
+        in the text.
+        """
+        text = format_text(text)
+        # we're bucketing based on the number of gendered words
+        f_count, m_count, n_count = get_finegrained_count(text, self.gender_dict)
+        # get f bucket
+        if f_count == 0:
+            f_bucket = 'f0'
+        else:
+            f_bucket = 'f1'
+
+        # get m bucket
+        if m_count == 0:
+            m_bucket = 'm0'
+        else:
+            m_bucket = 'm1'
+
+        # combine bucket
+        all_bucket = ' '.join([f_bucket, m_bucket])
+        return all_bucket
+
+    def _flip_ex(self, ex):
+        """
+        Return the counterfactual example for a given example (i.e. swap 'he' --> 'she')
+        """
+
+        def flip_str(txt_lines):
+            new_lines = []
+            lines = txt_lines.split('\n')
+            for text in lines:
+                f_text = format_text(text)
+                f_text_lst = f_text.split(' ')
+                new_words = []
+                for word in f_text_lst:
+                    if word in self.swap_dct:
+                        if word == 'her':
+                            # silly heuristic: choose him/his 50% of the time
+                            random_choice = random.choice([0, 1])
+                            if random_choice:
+                                new_word = 'his'
+                            else:
+                                new_word = 'him'
+                        else:
+                            new_word = self.swap_dct[word]['word']
+                    else:
+                        new_word = word
+                    new_words.append(new_word)
+                new_f_text = ' '.join(new_words)
+                uf_text = unformat_text(new_f_text)
+                new_lines.append(uf_text)
+            return '\n'.join(new_lines)
+
+        new_ex = deepcopy(ex)
```

Comment:
you can probably get away with a shallow copy don't you think? you're not mutating any lists